params/governance_paramset_test.go

```diff
@@ -0,0 +1,197 @@
+package params
+
+import (
+	"encoding/json"
+	"math/big"
+	"testing"
+
+	"github.com/klaytn/klaytn/common"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestGovParamSet_ParseValue(t *testing.T) {
+	zeroAddr := common.HexToAddress("0x0000000000000000000000000000000000000000")
+	mintingAmount := "9600000000000000000"
+	mintingAmountBig, _ := new(big.Int).SetString(mintingAmount, 10)
+
+	testcases := []struct {
+		ty     *govParamType
+		value  interface{}
+		parsed interface{} // If ok, expected value. Ignored if not ok.
+		ok     bool        // Expected 'ok'
+	}{
+		{govParamTypeGovMode, "none", "none", true},
+		{govParamTypeGovMode, "single", "single", true},
+		{govParamTypeGovMode, "ballot", "ballot", true},
+		{govParamTypeGovMode, "asdf", nil, false},
+		{govParamTypeGovMode, "", nil, false},
+		{govParamTypeGovMode, 1, nil, false},
+
+		{govParamTypeAddress, zeroAddr.Hex(), zeroAddr, true},
+		{govParamTypeAddress, zeroAddr, zeroAddr, true},
+		{govParamTypeAddress, 1, nil, false},
+
+		{govParamTypeUint64, int(7), uint64(7), true},
+		{govParamTypeUint64, uint(7), uint64(7), true},
+		{govParamTypeUint64, uint64(7), uint64(7), true},
+		{govParamTypeUint64, float64(1e9), uint64(1e9), true},
+		{govParamTypeUint64, "123", nil, false},
+		{govParamTypeUint64, -1, nil, false},
+		{govParamTypeUint64, -12.0, nil, false},
+
+		{govParamTypeBigInt, mintingAmount, mintingAmount, true},
+		{govParamTypeBigInt, mintingAmountBig, mintingAmount, true},
+		{govParamTypeBigInt, "123", "123", true},
+		{govParamTypeBigInt, "-123", nil, false},
+		{govParamTypeBigInt, "abc", nil, false},
+		{govParamTypeBigInt, "", nil, false},
+
+		{govParamTypeRatio, "100/0/0", "100/0/0", true},
+		{govParamTypeRatio, "10/20/30/40", nil, false},
+		{govParamTypeRatio, "", nil, false},
+
+		{govParamTypeBool, true, true, true},
+		{govParamTypeBool, 0, nil, false},
+		{govParamTypeBool, "", nil, false},
+	}
+
+	for _, tc := range testcases {
+		parsed, ok := tc.ty.ParseValue(tc.value)
+		assert.Equal(t, tc.ok, ok)
+		if ok {
+			assert.Equal(t, tc.parsed, parsed)
+		}
+	}
+}
+
+func TestGovParamSet_ParseBytes(t *testing.T) {
+	zeroAddrHex := "0x0000000000000000000000000000000000000000"
+	zeroAddr := common.HexToAddress(zeroAddrHex)
+	mintingAmount := "9600000000000000000"
+	mintingAmountBig, _ := new(big.Int).SetString(mintingAmount, 10)
+
+	testcases := []struct {
+		ty     *govParamType
+		bytes  []byte
+		parsed interface{} // If ok, expected value. Ignored if not ok.
+		ok     bool        // Expected 'ok'
+	}{
+		{govParamTypeGovMode, []byte("single"), "single", true},
+		{govParamTypeGovMode, []byte(""), nil, false},
+
+		{govParamTypeAddress, zeroAddr.Bytes(), zeroAddr, true},
+		{govParamTypeAddress, []byte(zeroAddr.Hex()), nil, false},
+		{govParamTypeAddress, []byte(""), nil, false},
+
+		{govParamTypeUint64, []byte{0x12, 0x34}, uint64(0x1234), true},
+		{govParamTypeUint64, []byte{}, uint64(0), true},
```

Comment:
What makes an empty slice ok when another case (line 101) is not?

Reply:
- I assumed that there could be some case where `0` is stored as an empty array in solidity.
That could happen because `big.NewInt(0).Bytes() == []byte{}`.

- For boolean, I wanted to be more strict so that only `0x00` and `0x01` are valid values.
This prevents accidently plugging in non-boolean values to boolean parameter.

Reply:
However, we can choose to interpret all truthy-values to true, falsey-values to value. i.e.

```go
govParamTypeBool = &govParamType{
    parseBytes = func(b []byte) (interface{}, bool) {
        return new(big.Int).SetBytes(b).Cmp(common.Big0) != 0, true
    }
}
```

@ian0371 What do you think about this change?

Reply:
I see. The strict policy would be better.